content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,82 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+
+- **Avoid using a reference solution at all if possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary: test cases can be generated with the answer known upfront, which eliminates an entire class of potential problems associated with the use of reference solutions:
+  - Extra time spent on running a reference solution
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution affecting the input passed to the reference solution
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function.
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s). Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
+
+
+## Random Tests
+
+Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
+
+- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
```

Comment:
Random tests primarily are there to prevent cheating. If there's a small number of edge cases or if there's a requirement to do something in case of, let's say, negative numbers, random tests don't add much to the fixed tests. Of course, if a solutions returns `None` only for `-42` instead of all negative numbers, it's not really correct, but checking for `-42` is as easy as checking for any negative number, so if for some reason adding negative numbers to the random tests is more difficult than just extending the range, it may not be worth the effort.

Reply:
You are right, I think I meant "all types of inputs" here, and not specifically edge cases. For example, when task deals with fractions, make sure that random tests generate various types of (allowed) fractions: proper fractions, reducible fractions, fractions with `num` > `den`, etc. They should not hope that randomness will provide all possible cases.

Maybe a suggestion that "one/few instances of an edge case could be shuffled into the input set if needed" cold be useful? Or not?

Reply:
Yes, maybe something like this.

Reply:
```suggestion
- **Each requirement should have a couple of corresponding random test cases.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate all possible scenarios by chance. If solutions are required to handle any edge cases, one or a few instances of each such case can be shuffled into the input set.
```

Sounds OK?

Reply:
Edge cases can be single values, like there's no point to test for `0!` in the random tests if it's already in the fixed tests.

Reply:
It generally should be in fixed tests, so maybe I should just remove this remark? Idk

Reply:
For an average kata, complexity of random tests should be proportional to complexity of bypassing fixed tests and to potential reward obtained by those who do so (the rank of the kata). As random tests should normally be present when it's not possible to cover the whole input range by fixed tests, the random tests should try generating special subsets of input values if they are not fully covered by the fixed tests. Probably something like that.

Reply:
```suggestion
- Random tests can be limited only to types of inputs, which are not covered with sample tests in the whole range. They should generate test cases for all scenarios which cannot be completely tested with fixed tests.
```

Maybe like this?

I am not sure how to go with a note on difficulty of working around them, because I am afraid that authors will be like "it's a 7 kyu kata, so it does not have to have good tests". What do you think?

Reply:
This looks OK, but I suppose "sample tests" here is a mistake?

Reply:
`/s/sample/fixed/`. thanks for spotting.